components/raftstore/src/store/peer_storage.rs

```diff
@@ -767,12 +799,191 @@ where
         Ok(())
     }
 
+    pub fn set_async_fetch_res(&mut self, res: Option<RaftlogFetchResult>) {
+        let in_fetching = match self.async_fetch_res.borrow().as_ref() {
+            r @ Some(RaftlogFetchResult::Fetched { .. }) => {
+                warn!(
+                    "unconsumed async fetch res";
+                    "region_id" => self.region.get_id(),
+                    "peer_id" => self.peer_id,
+                    "res" => ?r,
+                );
+                false
+            }
+            Some(RaftlogFetchResult::Fetching) => true,
+            None => false,
+        };
+        if res.is_some() {
+            self.async_fetch_res.replace(res);
+        } else if res.is_none() && !in_fetching {
+            self.async_fetch_res.replace(None);
+        }
+    }
+
+    fn async_fetch(
+        &self,
+        region_id: u64,
+        low: u64,
+        mut high: u64,
+        max_size: u64,
+        context: GetEntriesContext,
+        buf: &mut Vec<Entry>,
+    ) -> raft::Result<usize> {
+        match *self.async_fetch_res.borrow() {
+            Some(RaftlogFetchResult::Fetching) => {
+                // already an async fetch in flight
+                return Err(raft::Error::Store(
+                    raft::StorageError::LogTemporarilyUnavailable,
+                ));
+            }
+            _ => {}
+        }
+        let tried_cnt = if let Some(RaftlogFetchResult::Fetched {
+            ents: res_ents,
+            low: res_low,
+            max_size: res_max_size,
+            hit_size_limit: res_hit_size_limit,
+            tried_cnt: res_tried_cnt,
+            term: res_term,
+        }) = self.async_fetch_res.replace(None)
+        {
+            if res_low == low && res_term == self.hard_state().get_term() {
```

Comment:
Though it may be OK to use term of storage, using term from raft directly seems a safer choice.